operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java

```diff
@@ -3,29 +3,49 @@
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.javaoperatorsdk.operator.api.reconciler.Context;
 
-public abstract class AbstractDependentResource<R, P extends HasMetadata, C>
+public abstract class AbstractDependentResource<R, P extends HasMetadata>
     implements DependentResource<R, P> {
 
+  private final Creator<R, P> creator;
+  private final Updater<R, P> updater;
+  private final Matcher<R> matcher;
+
+  @SuppressWarnings("unchecked")
+  public AbstractDependentResource() {
+    creator = this instanceof Creator ? (Creator<R, P>) this : Creator.NOOP;
+    updater = this instanceof Updater ? (Updater<R, P>) this : Updater.NOOP;
+    matcher = this instanceof Matcher ? (Matcher<R>) this : Matcher.DEFAULT;
```

Comment:
I'm not sure about the matcher. It's existence is never checked, and this is a very generic problem. I would rather have it as a method with defaul implementation that can be overriden. Also in general does not fit in this CRUD picture of interfaces.

Reply:
Yes, that could be an other way to have it nicer.
But again matching is quite a general problem, that can be solved out of the box, for most cases (at least for K8S).

Reply:
Except that people might want to avoid using the rather inefficient generic implementation. Some of the examples actually override the method to use an optimised version. I've put the `match` method on `Updater` to make the link between both explicit. I've kept the `Matcher` interface to be able to provide standalone implementations that can be reused, in particular in `GenericKubernetesResourceMatcher`.